cranelift-wasm/src/code_translator.rs

```diff
@@ -1154,6 +1150,8 @@ fn translate_unreachable_operator(
                     builder.change_jump_destination(branch_inst, else_ebb);
                     builder.seal_block(else_ebb);
                     builder.switch_to_block(else_ebb);
+                    let params = params.clone();
+                    state.pushn(&params);
```

Comment:
This is only called for unreachable code, so I'm not concerned about performance here, but it looks odd; why is a clone needed if you're just doing a borrow?

Reply:
It's because `params` is borrowed immutably from within `state`, but calling `state.pushn` requires a `&mut self` which isn't possible while `params` is borrowed. Internally, it would be safe, since they operate on disjoint fields of `TranslationState`, but since the borrow checker can't see through method calls, it doesn't realize this.

There have been proposals for ways to work around this failing of the borrow checker, but nothing accepted and implemented yet. The only current work arounds are to either inline all the methods by hand so that the borrow checker sees everything, or make these things free functions that take the fields they operate on as parameters, rather than borrowing every field `&mut self`.